contracts/modules/Experimental/Checkpoints/WeightedVoteCheckpoint.sol

```diff
@@ -0,0 +1,188 @@
+     * @param _polyToken Address of the polytoken
+     */
+    constructor(address _securityToken, address _polyToken)
+    public
+    Module(_securityToken, _polyToken)
+    {
+
+    }
+
+    /**
+     * @notice This function returns the signature of configure function
+     */
+    function getInitFunction() external pure returns(bytes4) {
+        return bytes4(0);
+    }
+
+    /**
+     * @notice Allows the token issuer to create a ballot
+     * @param _duration The duration of the voting period in seconds
+     * @param _noOfProposals Number of proposals
+     */
+    function createBallot(uint256 _duration, uint256 _noOfProposals) external {
+        _onlySecurityTokenOwner();
+        require(_duration > 0, "Incorrect ballot duration.");
+        uint256 checkpointId = ISecurityToken(securityToken).createCheckpoint();
+        uint256 endTime = now.add(_duration);
+        _createCustomBallot(now, endTime, checkpointId, _noOfProposals);
+    }
+
+    function _createCustomBallot(uint256 _startTime, uint256 _endTime, uint256 _checkpointId, uint256 _noOfProposals) internal {
+        require(_noOfProposals > 1, "Incorrect proposals no");
+        require(_endTime > _startTime, "Times are not valid");
+        uint256 ballotId = ballots.length;
+        uint256 supplyAtCheckpoint = ISecurityToken(securityToken).totalSupplyAt(_checkpointId);
+        ballots.push(
+            Ballot(
+                _checkpointId, supplyAtCheckpoint, _startTime, _endTime, 0, _noOfProposals, true
+            )
+        );
+        emit BallotCreated(_startTime, _endTime, ballotId, _checkpointId, _noOfProposals);
+    }
+
+    /**
+     * @notice Allows the token issuer to create a ballot with custom settings
+     * @param _startTime Start time of the voting period in Unix Epoch time
+     * @param _endTime End time of the voting period in Unix Epoch time
+     * @param _checkpointId Index of the checkpoint to use for token balances
+     * @param _noOfProposals Number of proposals
+     */
+    function createCustomBallot(uint256 _startTime, uint256 _endTime, uint256 _checkpointId, uint256 _noOfProposals) external {
+        _onlySecurityTokenOwner();
+        _createCustomBallot(_startTime, _endTime, _checkpointId, _noOfProposals);
+    }
+
+    /**
+     * @notice Allows a token holder to cast their vote on a specific ballot
+     * @param _ballotId The index of the target ballot
+     * @param _proposalId Id of the proposal which investor want to vote for proposal
+     */
+    function castVote(uint256 _ballotId, uint256 _proposalId) external {
+        require(_ballotId < ballots.length, "Incorrect ballot Id");
+        Ballot storage ballot = ballots[_ballotId];
+
+        uint256 weight = ISecurityToken(securityToken).balanceOfAt(msg.sender, ballot.checkpointId);
+        require(weight > 0, "weight should be > 0");
+        require(ballot.totalProposals >= _proposalId && _proposalId > 0, "Incorrect proposals Id");
```

Comment:
Maybe we should allow voting for proposal Id 0 and consider that as reject vote. i.e the voter is against all the proposals.

Reply:
I think we should allow voting for proposal 0 - not sure why we wouldn't index proposals from 0? I don't think proposal 0 should have any specific special meaning though.